### ISSUE TITLE
[REVIEW]: stacomiR : a common tool for monitoring fish migration

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#791]
**Submitting author:** @MarionLegrandLogrami (<a href="http://orcid.org/0000-0002-4688-8303">Marion Legrand</a>)
**Repository:** <a href="https://github.com/MarionLegrandLogrami/stacomiR" target ="_blank">https://github.com/MarionLegrandLogrami/stacomiR</a>
**Version:** 0.5.3.1
**Editor:** @arfon
**Reviewers:** @steffilazerte, @boshek
**Archive:** <a href="https://doi.org/10.5281/zenodo.3371259" target="_blank">10.5281/zenodo.3371259</a>

## Status

[![status](http://joss.theoj.org/papers/0cacbf19f5822f668e915f788dec1a82/status.svg)](http://joss.theoj.org/papers/0cacbf19f5822f668e915f788dec1a82)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/0cacbf19f5822f668e915f788dec1a82"><img src="http://joss.theoj.org/papers/0cacbf19f5822f668e915f788dec1a82/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/0cacbf19f5822f668e915f788dec1a82/status.svg)](http://joss.theoj.org/papers/0cacbf19f5822f668e915f788dec1a82)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/MarionLegrandLogrami/stacomiR" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@steffilazerte, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @arfon know.

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/MarionLegrandLogrami/stacomiR">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (0.5.3.1)?
- [x] **Authorship:** Has the submitting author (@MarionLegrandLogrami) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon. I'm here to help you with some common editorial tasks. @steffilazerte it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[--> Check article proof :page_facing_up: <--](https://github.com/openjournals/joss-papers/blob/joss.00791/joss.00791/10.21105.joss.00791.pdf)

**Comment by MarionLegrandLogrami** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #791 with the following error: 

 /app/vendor/bundle/ruby/2.3.0/bundler/gems/whedon-195e6d124d0f/lib/whedon.rb:80:in `check_fields': Paper YAML header is missing expected fields: tags (RuntimeError)
	from /app/vendor/bundle/ruby/2.3.0/bundler/gems/whedon-195e6d124d0f/lib/whedon.rb:68:in `initialize'
	from /app/vendor/bundle/ruby/2.3.0/bundler/gems/whedon-195e6d124d0f/lib/whedon/processor.rb:32:in `new'
	from /app/vendor/bundle/ruby/2.3.0/bundler/gems/whedon-195e6d124d0f/lib/whedon/processor.rb:32:in `set_paper'
	from /app/vendor/bundle/ruby/2.3.0/bundler/gems/whedon-195e6d124d0f/bin/whedon:37:in `prepare'
	from /app/vendor/bundle/ruby/2.3.0/gems/thor-0.20.0/lib/thor/command.rb:27:in `run'
	from /app/vendor/bundle/ruby/2.3.0/gems/thor-0.20.0/lib/thor/invocation.rb:126:in `invoke_command'
	from /app/vendor/bundle/ruby/2.3.0/gems/thor-0.20.0/lib/thor.rb:387:in `dispatch'
	from /app/vendor/bundle/ruby/2.3.0/gems/thor-0.20.0/lib/thor/base.rb:466:in `start'
	from /app/vendor/bundle/ruby/2.3.0/bundler/gems/whedon-195e6d124d0f/bin/whedon:99:in `<top (required)>'
	from /app/vendor/bundle/ruby/2.3.0/bin/whedon:22:in `load'
	from /app/vendor/bundle/ruby/2.3.0/bin/whedon:22:in `<main>'

**Comment by MarionLegrandLogrami** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[--> Check article proof :page_facing_up: <--](https://github.com/openjournals/joss-papers/blob/joss.00791/joss.00791/10.21105.joss.00791.pdf)

**Comment by MarionLegrandLogrami** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[--> Check article proof :page_facing_up: <--](https://github.com/openjournals/joss-papers/blob/joss.00791/joss.00791/10.21105.joss.00791.pdf)

**Comment by steffilazerte** [comment]
The premis of `stacomiR` is to provide access to data base for monitoring fish migration in a user-friendly way. This is a great idea! I think it's really important to provide access to large biological data sets, especially in a way that is inclusive and which helps users who aren't familiar with R. The amount of work that has gone into this R package is also substantial and I'm eager to see it in action.

Unfortunately, I ran into several major issues when I tried to install and use `stacomiR`. 

1) Lack of documentation
While the functions themselves are well documented, there are no tutorials, vignettes or websites with information for users just getting started. Even the documentation revealed by `?stacomiR-package` didn't provide any information for a new user, but merely directed the user to contact the author.

2) Installation
Dependencies in the DESCRIPTION don't list minimum versions. This may create problems for some users with older package versions. For example, I ran into a problem with the installation because I have Hmisc v4.0.3, but the package required Hmisc >= v4.1.0

3) Usage
Possibly because of both the above, once installed I struggled to use the package.

### Specific problems 

I followed the instructions in the JOSS paper, but was unable to launch the interface with `stacomi()`, or with `stacomi(gr_interface = FALSE, login_window = FALSE, database_expected = FALSE)`.

I expected the first command to fail as I don't have the database (and it did) but the second command did nothing at all (no output, no errors), and I was at a loss of how to further proceed.

Following the rest of the example I ran into an error with the `calcule()` function:

```
> r_mig_mult <- calcule(r_mig_mult, silent = TRUE)
Loading required package: DBI
Error in postgresqlNewConnection(drv, ...) : 
  RS-DBI driver: (could not connect test@localhost:5432 on dbname "test": could not connect to server: Connection refused
	Is the server running on host "localhost" (127.0.0.1) and accepting
	TCP/IP connections on port 5432?
)
Error in !dbPreExists : invalid argument type
```

The final lines in the JOSS paper gave me no output (presumably because the above function failed). I also struggled to understand what the functions were supposed to do.

### Other issues:

- Authorship: The package is authored by three people, but the JOSS paper only by two. Also the JOSS paper's first author seems to have made the fewest contributions (only judging by the number of Adds/Commits in the subversion repository). Perhaps the authors could comment on this.

- In the JOSS paper, the repository is listed as GitHub, but the GitHub repository isn't a complete R package (it's missing DESCRIPTION file, for one). It also looks as if the main source is located on R Forge. I would suggest either updating the GitHub repository to be instep with R Forge, or omitting the GitHub repository completely and just stick with R Forge.

- From what I understand of JOSS policies, there should actually be very little documentation in the JOSS paper. This paper should just be a summary of what the program is capable of, not information on how to use it.

- Which regions are included in this database? France? Europe? etc.?

### Verdict
In the end, I can't say I've actually reviewed the package, as I wasn't able to get it working. However, dealing with remote databases and interactive interfaces are always tricky. I'm still looking forward to giving it a shot and I'm more than happy to work with the authors to figure out what's going on. 

For acceptance in JOSS I definitely need to finish my review (i.e. be able to test the package), and there also needs to be some solid documentation (perhaps a package vignette?) that walks new users through getting set up and addresses common problems or road blocks.


### My System Info

```
 version  R version 3.4.4 (2018-03-15)  
 os       Ubuntu 16.04.4 LTS            
 system   x86_64, linux-gnu           
 ui       RStudio                     
 language en_CA:en                    
 collate  en_CA.UTF-8                 
 tz       America/Winnipeg            
 date     2018-06-28 
```

**Comment by arfon** [comment]
Thanks for your details review @steffilazerte. @MarionLegrandLogrami  - please take a look at @steffilazerte's feedback.

**Comment by MarionLegrandLogrami** [comment]
Dear ​Steffi LaZerte,

First of all, Cédric Briand and I wanted to warmly thank you for the time you spent to test our package. We found your remarks really useful and some will allow us to improve it​​.
I will take your remarks/questions one by one and in the order of your message to try to answer as best as possible.

**1. Lack of documentation** 
Indeed, the documentation of the R package does not provide any help on how to use the package itself (link to the database, the installations to be made, etc.). To present in more details the package we wrote a vignette but we failed to compile and integrate it into the package. We did not understand why, but one of your remarks enlightened us on this point (I will come back to this later). Until we can integrate this vignette to our package you can already find it at this address : https://github.com/MarionLegrandLogrami/stacomiR/tree/master/vignettes.
In addition, we created a space dedicated to this project on "trac" system. This space has been created for users of the stacomi project. They can find different elements to help them to better use this tool. You can access to this space following this link : http://w3.eptb-vilaine.fr:8080/tracstacomi/wiki. You will need to connect to this space using user (utilisateur) : stacomi and password (mot de passe) : stacomi.
To give you some background on the fish data banking in France, you should know that there is no centralized system to organize this collection. The data are collected by various associations whose purpose is to follow fish on the long term (especially diadromous fish). Sometimes this is done by the Amateur Fishing Federations and the Migratory Associations (10 in France that work on diadromous fish) of the watershed concerned, pool the informations of the different Federations and make sure that the protocols are as homogeneous as possible. Sometimes it is the Migratory Associations themselves who collect the data (the most frequent case). Finally, there may be public institutions or research laboratories that monitor stations. To date, there is no structure to homogenize this collection or to organize the valuation of these data. It is on this basis that we decided to develop a R package to enable all data producers, whatever the structure, to improve the banking and the valuation of their data and if wanted to share data. Data ownership obviously remains that of producers and only producers can decide whether or not they want to make their data available and in what context. However, the R package is an excellent tool to show data producers the value of sharing information to compare data across multiple stations, multiple watersheds, and the group of Stacomi users that we created Cédric and I allows them to meet each other, share their package practice and make sharing information easier. Ultimately, we want to develop a package using Shiny and that would go further in the valuation and sharing of information.The stacomi project allows to make available a common tool and Cédric and I provide assistance to users concerning the installation of the tool, the difficulties of use, etc.

**2. Installation**
It's a good point, thank you for highlighting it. On rforge website, we indicated that to work the package needs to be run with r 3.5.1 in particular since some functions of the new version of Hmisc don't work anymore with previous version of R. But, it would be better to indicate the version of dependencies in the description as you propose. I wrote a ticket on the trac to remember this improvement for a new version of the package ([ticket number 238](http://w3.eptb-vilaine.fr:8080/tracstacomi/ticket/238)).

**3. Usage**
We hope that all clarifications gave in the two first point will help !

**4. Specific problems**
The problem you submitted to us regarding the impossibility of connecting to the "test" database gave us an interesting hint to understand why our vignette did not compile. For some calculations that we do in the package we have created a database named "test" that has been created for all users. Nevertheless, a person who does not have the stacomi installation obviously does not have a test database. We will work on a new solution to no longer need this database. I created a ticket on the trac corresponding to this improvement ([ticket number 239](http://w3.eptb-vilaine.fr:8080/tracstacomi/ticket/239)).
The commande line  stacomi() works only for users having a complete installation as you guessed. The command line  stacomi(gr_interface = FALSE, login_window = FALSE, database_expected = FALSE) allows you to use the package but only in r command line as the interface need a direct connexion to the database. The reason why the next command didn't work (r_mig_mult <- calcule(r_mig_mult, silent = TRUE)) is because we ask in the calcule method of this function to connect to the "test" database. This is a mistake and it is the reason I created a trac ticket ([239](http://w3.eptb-vilaine.fr:8080/tracstacomi/ticket/239)) to improve the next version of our package. To work around this problem immediately, is it possible to you to create a localhost with UTF8 encoding database under postgresql named "test" (landlord = postgres) and a test login role with user=test and password=test. This will allow the command line  r_mig_mult <- calcule(r_mig_mult, silent = TRUE)  to work.

**5. Other issues**

- Authorship: I added Timothée Besse in the author of the paper as he helps us to develop some parts of the package. For the order of the name in the paper, the reasons why we chose to put my name first even if I made much less commit than Cédric on the package are the following: while Cédric worked mainly on the development of the R package, I collected data on fish migration from more than 20 different structures in order to bank all the data in the postgresql database. This bankarization (more than 18 million of fish passage data) allowed the data producers to use the package's functionalities on all their data (including historical data) which was essential for the tool to become interesting for them. Thus, we divided the work mainly between Cédric and I to be the most effective. Of course we worked collaboratively which allowed everyone to follow the progress of the other and when necessary to give some help. On the other hand, I carry out a thesis on the modification of fish migration timing in France which required the use of the collected data (article being re-read by co-authors). It seemed to us more interesting to put my name first to value all this work in the context of my thesis (this choice is obviously shared by Cédric and I). To give you an idea of the work, here is a map of the monitoring stations whose data I banked. The oldest stations have a time series of more than 30 years

![stacomi_localisation_ac_conv_2018_03_23](https://user-images.githubusercontent.com/26621584/42285905-3ac40928-7fb1-11e8-9c3f-b7aa06a71351.png)

- Repository: Indeed, all the package is developped on rforge. But we thought that to submit a JOSS paper we had to have a github repository. That's why we copied / pasted the different files into the github directory created specifically to submit the paper to JOSS. As we didn't know how to change now the information for the directory, I update the github repository to add DESCRIPTION file.

- JOSS policies: Does this mean that you recommend to delete the different lines of R code we write on the paper to give some illustration on the possibilities of the package? [We saw for example this paper published on JOSS](https://joss.theoj.org/papers/87b007aadfc9f96dfff5fc001b34a2b0) which incorporated command lines of their package R and we thus thought that it was allowed. But if you think it would be better to delete them to give a better overview of what the package can do we are not opposed. 

- Region included: at the moment the database is used all over France (thanks in large part to the banking work I have done for my thesis) and also in Spain. We have planned that the package can be used quite widely and that is why we have set up a system for translating the messages sent by the package in French and English for all messages and in Spanish for only a part (translation system realized with po.edit). To go further and promote the package in other countries, we should be able to put more resources on the translation, but if one day it becomes necessary, everything is planned so that we can do it.

We hope that these informations will be useful to you. Do not hesitate to come back to us if there are any questions or if you encounter difficulties.
Thank you again for the time invested, which was very helpful to us.

**Comment by MarionLegrandLogrami** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[--> Check article proof :page_facing_up: <--](https://github.com/openjournals/joss-papers/blob/joss.00791/joss.00791/10.21105.joss.00791.pdf)

**Comment by arfon** [comment]
Thanks for the detailed review @steffilazerte. Responding to a couple of your comments here:

> This is good, but for an open source project, it might be better to have an open tracking system as well. At the very least there should be some information for users as to where they can go to submit issues. Consider using the "BugReports: " tag in the DESCRIPTION (for example see here)

This is actually a requirement for JOSS submissions (see https://joss.readthedocs.io/en/latest/submitting.html#submission-requirements):

> ...the software associated with your submission must:
> - Have an issue tracker that is readable without registration

> Thank you for the clarification. I was sure there was a rational, but part of my job as a reviewer is checking up on this. @arfon, do you think it could be useful to have authors declare authorship contributions as part of the review application? It's not very common, but I know of at least some journals which require this.

We're considering this as an editorial board but at this point we don't have a policy for this.

**Comment by arfon** [comment]
Hi @MarionLegrandLogrami - when do you think you will be able to incorporate @steffilazerte's feedback?

**Comment by MarionLegrandLogrami** [comment]
Hi @arfon and @steffilazerte  - thanks for your message. A very short message just to tell you that we are incorporating the @steffilazerte's remarks. it's hard to tell you when we're done but we've made good progress. I will send you a message as soon as the package is ready and in accordance with the requests made

**Comment by MarionLegrandLogrami** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.00791/joss.00791/10.21105.joss.00791.pdf)

**Comment by MarionLegrandLogrami** [comment]
Hi @arfon and @steffilazerte - I'm happy to inform you that we finished to update the stacomiR package. The new version is available on CRAN and rforge (we updated the version provided on github as well). We made the main following modifications: 

- remove all call to test database, 

- incorporate the vignette inside the package, 

- add [detailed installation instructions on github](https://github.com/MarionLegrandLogrami/stacomiR/blob/master/Installation/README.md) for new user, 

- provide a database with one example schema (with data) and one empty schema (for people who want to use the database to organize their data),

- add [contribution file](https://github.com/MarionLegrandLogrami/stacomiR/blob/master/CONTRIBUTING.md) and a BugReports part inside the [description file](https://github.com/MarionLegrandLogrami/stacomiR/blob/master/DESCRIPTION), 

- update the paper we submitted to add some context informations and add a link to the vignette. 

We hope that this new version will meet all your requirements.
Again we wanted to thank @steffilazerte for all the really helpful comments that helped us to improve our package.

**Comment by arfon** [comment]
Thanks @MarionLegrandLogrami! @steffilazerte - please take another look when you get a chance.

**Comment by steffilazerte** [comment]
I will definitely take a look, but it might be a week or so before I can really dive in fully. I'll keep you posted.

**Comment by steffilazerte** [comment]
@MarionLegrandLogrami and coauthors, 

I'm happy to say that with the updates to the `stacomiR` package I've been able to run through the examples based on the built-in example data sets!

The vignette is a wealth of information and a valuable addition. By running through the example data, the plot/report documentation and checking back with the vignette's section on Data structure, I was able to gain a much better understanding of the kind of data being summarized (which is impressive, given how little I know about fish monitoring).

I was able to snoop through most of the documentation and create many different graphical summaries of the included data sets. I explored figures and summaries for these data sets and, once again, I am seriously impressed by the variety of analyses included.

Including instructions for installing/restoring the database on GitHub is a great idea, and I like that you've linked to them in the JOSS manuscript as well.

The downside is that I am still unable to setup the database, and I haven't been able to launch the user interface.

I've also realized that your setup is designed for Windows (whereas I am on Linux). For example, I will never be able to have the location `c:\users\my_session\Documents\CalcmigData` on my computer, as Linux file paths do not have drive locations (`c:\`) (and neither do Macs).

Below are the error messages I got when attempting to restore the database. I suspect that the `locale` error is related to the Operating system differences, but I am not sure if the `database "bd_contmig_nat" does not exist` error is the result of the `locale` error or something else.

```
postgres@steffi-desktop:~$ psql postgres < install_bd_contmig_nat.sql 
SET
SET
SET
SET
SET
 set_config 
------------
 
(1 row)

SET
SET
SET
ERROR:  invalid locale name: "French_France.1252"
ERROR:  database "bd_contmig_nat" does not exist
\connect: FATAL:  database "bd_contmig_nat" does not exist
postgres@steffi-desktop:~$
```

### Suggestions

- In the installation instructions, add Windows as a system requirement

- In my quest to access the database, I solved at least one error by updating my PostgreSQL to version 9.6, you may want to add this version requirement to the installation instructions.

- Be **very** explicit about the necessity of having a database in the vignette. I would suggest right at the top, under the "Usage" section, that you state that the following two commands will only work once you've installed (or created) a database. Then link to the installation instructions on GitHub.

- In the vignette example with silver eels and Durif stages, should it be plot.type = 1? or plot.type = 2? (I'm definitely not an expert, but expected to see the word 'Durif' somewhere on the figure)

### Conclusions

@arfon, I have evaluated everything I can here. The package authors have addressed my comments and concerns and I have run through all the JOSS review check boxes, with the exception of fully testing the functionality (i.e. accessing the database in order to create reports of novel data subsets and the graphical user interface). Admittedly, these aspects are a pretty big part of this R package.

I would suggest either accepting the revisions as sufficient, or finding another, windows-based reviewer to finish the review. My apologies for not being able to take this to the end. I hope this hasn't been too inconvenient for everyone involved!

**Comment by arfon** [comment]
> @arfon, I have evaluated everything I can here. The package authors have addressed my comments and concerns and I have run through all the JOSS review check boxes, with the exception of fully testing the functionality (i.e. accessing the database in order to create reports of novel data subsets and the graphical user interface). Admittedly, these aspects are a pretty big part of this R package.

Thanks for your excellent review @steffilazerte and your suggestion here. I think we should seek an additional reviewer to help verify these aspects of the package.

@MarionLegrandLogrami - do you know of anyone who might be able to help with this? @karthik - do you know of any Windows/R users that might be able to help?

**Comment by karthik** [comment]
@arfon Sure, let me come up with some ideas

(apologies about ooo. Trying to figure out how to unset the test date).

**Comment by MarionLegrandLogrami** [comment]
@steffilazerte, my colleagues and I wanted to thank you once again for all the work done and the help provided. Thanks to you we made lot of progress ! For the error concerning the database 

> database "bd_contmig_nat" does not exist
I think this is not related to the operating system. I saved the database using the pg_dump command with the -C option to insert in the dump file the command line needed to create the database but unfortunately (I do not understand why and I did not see this problem before reading your message) the psql command fails to restore the database without the user manually creating the database using the following SQL command line:
> CREATE DATABASE bd_contmig_nat WITH ENCODING = 'UTF8';
I will add this command line in the github instruction file for the installation of the database.

Thanks again for all your work and all your remarks and hints.

**Comment by steffilazerte** [comment]
Oh that's useful, I'm glad I was able to help at least a little! Good luck!

**Comment by MarionLegrandLogrami** [comment]
@karthik Hello, thank you for your help in trying to find an additionnal reviewer. Do you have some suggestions ?
@arfon I'm sorry but I don't know who else to propose to help finish the job

**Comment by arfon** [comment]
Friendly reminder to give us some suggestions here @karthik if you have any?

**Comment by karthik** [comment]
I'd recommend @boshek if he's got time.

**Comment by arfon** [comment]
:wave: @boshek - do you think you might be able to help us here? We're very close to finishing this review but unfortunately @steffilazerte wasn't able to verify some of the functionality of the software.

**Comment by boshek** [comment]
Rather than test on Windows which so far I have unable to accomplish I'd suggest two steps that would make this package more usable:
- Move the workflow of steps from a click based approach to R. For example, something like this:
> You need to download the stacomi.zip file, unzip the file and put the stacomi folder at the root of your c:\Program Files repository. You will find 6 files inside the stacomi folder

could likely be turned into a function using some combination of `download.file` and `unzip`.  Another example of this is the sql used to create the database. I believe some of this can directly be run from R (as well as the above download/unzip workflow described above). 

- More importantly though I think this package should very strongly consider adapting code so that it can be run on multiple platforms. If broad adoption is a goal here, restricting to Windows seems like a mistake. Fortunately, the `rappdirs` package provides a very nice option of platform-independent locations to download data (i.e. the content of stacomi.zip). So that is you'd replace something like this:

```
filecsv<-"C:/Program Files/stacomi/calcmig.csv"
```

with this: 

```
> filecsv <- file.path(rappdirs::site_data_dir("stacomiR"), "calcmig.csv")
> filecsv
[1] "C:\\PROGRA~3\\stacomiR\\stacomiR/calcmig.csv"
```

**Comment by MarionLegrandLogrami** [comment]
Hi @boshek . Thanks a lot for your comments. 
First of all, I wanted to tell you that we are looking at your proposal to simplify the installation process, but we need some time to do it. I will come back to you when it's over
Regarding the use of multiple platforms, the stacomiR package [is already working on linux](https://r-forge.r-project.org/R/?group_id=1019). We have not checked the use of the package under Linux because we are all working on Windows, but we will consider configuring a virtual machine with Linux to test it. At this moment, all the users of this package are using windows operating system, except one that is using OS Mac. All these users are not very familiar with R (most have never used R) no more than with databases and ODBC link creation. For all these users we made all the installation for them. But we agree that trying to find a way to simplify this installation would be a good thing.

To respond more specifically to some of your comments :

> You also need to add new connection role: [...]

You need to do this **only** if you want to use the database (= add data) and you want to change the name of the user's schema (I will try to explain that better on github). To simply test the package you don't need to do that as you have the IAV schema with all data to test the package.

It's the same for the Updating_grant.zip file. You must run the file only if you want to change the schema username (you just need to open the SQL script file in pgAdmin and execute the code).

Regarding the error you get:

> > stacomi()
Error in funout(msg3, arret = TRUE) : connection impossible
In addition: Warning messages:
1: In RODBC::odbcDriverConnect("DSN=bd_contmig_nat;UID=iav;PWD=iav",  :
  [RODBC] ERROR: state IM002, code 0, message [Microsoft][ODBC Driver Manager] Data source name not found and no default driver specified
2: In RODBC::odbcDriverConnect("DSN=bd_contmig_nat;UID=iav;PWD=iav",  :
  ODBC connection failed

- Do you have a PostgreSQL ODBC driver ? If you have an R 32 bits you need a PostgreSQL ODBC driver 32 bits (if R is 64 bits you need a 64 bits ODBC driver). 

- Did you configure a PostgreSQL ODBC link (Unicode), with the name of the database, the server and port, etc. ?

I'll be back to you shortly to inform you of our progress on the improvements you have submitted to us.

**Comment by labarba** [comment]
👋  @MarionLegrandLogrami — We haven't heard from you in a while. What's your status? Are you planning on improving the software in view of the feedback you have received here?

**Comment by MarionLegrandLogrami** [comment]
@labarba  - thank you for your message. We are still working on the change requested by our reviewer. We are almost done (still a few checks) and try to finish as soon as possible.

**Comment by MarionLegrandLogrami** [comment]
Hi @boshek and @labarba - We are happy to inform you that we completed the development of the automated installation of our package stacomiR and that we commited the scripts on github. We produced two versions of the automated installation: one for Windows and the other for Linux. To date, we thoroughly tested the installation under Windows. We still have some checks to perform on the Linux version. Indeed, we checked that the script creates the database, restores the data available on github and creates an ODBC link but we still have to test the launch of the package (in particular, check that the connection is well established with the database). We plan to do these final checks (for the Linux version) by the end of next week. We hope that this development will meet your expectations. We believe that it will significantly simplify the installation of our package and make it more easily accessible. We wanted to thank @boshek for this hint.

**Comment by MarionLegrandLogrami** [comment]
Hi @boshek - a little word just to let you know that we commited on github the final version of the automated installation for Linux. I forgot to tell you last time that we first tried to developp an R script to make all the installation (we tested the functions you suggested), but we failed. It's the reason we thought of an other way with batch files. We hope this will meet your expectations.

**Comment by arfon** [comment]
:wave: @boshek - could you please review the changes that have been made by @MarionLegrandLogrami and let us know if these satisfy your concerns?

**Comment by arfon** [comment]
Friendly reminder that we’re waiting on your input here @boshek :smile:

**Comment by boshek** [comment]
@arfon. Sorry this got missed on the last reminder for some reason. I will allot some time towards this.

**Comment by boshek** [comment]
Hello @MarionLegrandLogrami and @arfon 

It is possible that I am out of depth here. I acknowledge that. However if I am asked to provide an honest assessment here I don't think that the authors have addressed my concerns above. I am having  trouble getting the database set up on Windows. My original recommendation is that all the database set up be moved into R functions; that is wrap the shell script in an R function. In the absence of functions that do this the installation instruction need to be comprehensive - that is step by step. I am making the assumption here that the audience of this package is an average R user without any (or maybe only a little) knowledge of PostgreSQL. With that in mind this section:

https://github.com/MarionLegrandLogrami/stacomiR/tree/master/Installation#22-on-windows-host-

assumes too much knowledge on the users part. 

I am unable to call the `install_stacomi.cmd`. Perhaps there is additional setup, which may be obvious, but if I am the naive new user I don't think that I would be the only person to run into these issues. For example this line: 

> You WILL have to give a SQL user with admin rights (create a database, create a connection role). Else the install will fail.

assumes that a user even knows what this means or how to go about doing it. My assessment here might be reflective of my relatively poor understanding of PostgreSQL. However if the authors have a different audience for the package in mind, rather than an R user with limited knowledge of databases, perhaps this is fine. I wish to be clear in saying that this is not an indictment of the authors work. The skills and technical expertise is clear and impressive. However I am still not able to get this up an running using the current instructions.

**Comment by arfon** [comment]
Thanks for your honest feedback here @boshek - it's much appreciated.

@MarionLegrandLogrami - both of the reviewers of this submission have had serious challenges getting the database up and running for testing. While @boshek is being very gracious stating that they don't feel like an expert with these technologies I think their point that this is going to be very hard for _any_ users of your software is a very good one.

I'm not exactly sure what to suggest at this point but it's clear that the installation documentation is still lacking for new users. Perhaps you could find someone local to your host institution to try and follow the installation instructions so you can get a better understanding of what challenges people are running into.

As things stand, both reviewers have dedicated significant time and effort to these reviews but thus far, it's not been possible for either of our reviewers to fully verify the functionality of this software which means that this review is currently blocked.

**Comment by MarionLegrandLogrami** [comment]
Hi @boshek and @arfon,

The progam we provided to help to install everything is a cmd file. So it is an executable file. To launch it, you have to double-clik on the file. Doing this you will launch the program with some default settings. If you want to change these default settings you have to call the program in a console (start menu / cmd+enter to call the command) and enter the settings you need. I made some changes to the documentation in github to explain more clearly how to do this.
If you really want to execute this program inside R, you can call the shell function. I explained everything in the documentation:

#Inside R, if you want to launch the program with the default settings
scriptAddr<-"path_to_install_stacomi.cmd" (e.g. "C:/Users/Public/stacomir/Installation/install_stacomi.cmd")
shell(scriptAddr,"cmd")

#If you need to change some settings, you first have to look at the options provided by the install_stacomi.cmd program. To do this type:
scriptAddr<-"path_to_install_stacomi.cmd" (e.g. "C:/Users/Public/stacomir/Installation/install_stacomi.cmd")
command<-paste(scriptAddr,"-h")
shell(command,"cmd")

##To change some options:
scriptAddr<-"path_to_install_stacomi.cmd" (e.g. "C:/Users/Public/stacomir/Installation/install_stacomi.cmd")
command<-paste(scriptAddr,"-H hostname -P port -u user -p password")
#Example with hostname=localhost (=default settings), port=6543, user=foo, password=foofoo
command<-paste(scriptAddr,"-P 6543 -u foo -p foofoo") #I don't need to add -H localhost as it is the default settings
shell(command,"cmd")

I tried to clarify in the documentation the admin rights for the SQL user. If you made the postgreSQL installation yourself it is the username/password you provided during the installation. If you didn't install the program by yourself you need to ask the person who did it to give you a username with the rights to create a database and a connection role. Indeed to make graphs with the stacomiR using the graphical interface provided by the package you need to have data in a database. This database will be created and fill with a sample set of data automatically  with the cmd program but to do this you really have to have a username with the right to create a database. The user you create during the installation of PostgreSQL have all the rights needeed.

The difficulty with the installation is that we have to connect R with other programs like postgreSQL. In France and Spain, several structures use this program and as these users often have no knowledge neither in R nor in database, we provide a support of help to all the users by realizing ourselves the installation of everything that is needed on their computers.
Thanks to our wish to publish our package in JOSS and the comments of our 2 reviewers, we have done a lot of work to simplify the installation. But actually, the user must always have some knowledge including how PostgreSQL works if he wants to be autonomous in the installation (otherwise, he must contact us so we can help). We can hardly see how to do otherwise. Indeed, if the user does not know if postgreSQL is installed on his computer on a localhost or on a distant server or does not have in mind the user and/or the password provided during the installation of postgres this can cause difficulties during installation. In our cmd program we have entered default values ​​for the server (= localhost), for the port (= 5432 i.e. the default port on which postgresql is installed) and for user and password (postgres for both). But the user will certainly have to change at least the user and the password (except if he chose this user and password during the installation). This is possible either with the command prompt or now in R with the shell function.
We hope that all these informations will help.
@boshek: don't hesitate to contact us if there are still things that are not clear. We will be happy to help you. We spent lot of time trying to improve our package and we really hope to publish it in JOSS

**Comment by boshek** [comment]
Hi @MarionLegrandLogrami - just a quick check in here to ask if you've considered putting all this into a docker container which can reproducibly run. That was you'd only have to provide instructions to the user on how to use docker. How feasible do you think that would be?

**Comment by MarionLegrandLogrami** [comment]
Hi @boshek - Thank you for your message. We spent some time reading a little about the docker container with which we are not familiar. This technology seems very interesting, but we are not sure that it will facilitate the installation (as it seem to be not really easy to use docker especially if the goal is to run an R package with a graphical interface - how can you access the graphical interface and interact with it inside the docker is still a mistery to me). What about the shell() command we suggested to use to launch the installation inside R ? Have you had any difficulties ? Have you managed to install postgresql ? If you encounter difficulties with any part of the installation don't hesitate to contact us. We will be very happy to help you (for example by using Teamviewer to help you to install everything)

**Comment by boshek** [comment]
Hi @MarionLegrandLogrami - I still think that the whole script can be abstracted away into an R function and the script itself could be included in the package in `inst` folder. Arguments to think function could be the username, password, host and port. That function would run the script for you. I think if you asked users if they were ok with the function running an executable you'd be ok.

In terms of the process, I am having trouble with the odbc part of the installation. On my machine the 64 bit odbc driver causes a silent crash of pgAdmin and no erro message. Installation of the 32 bit driver proceeds but the program executable seems unable to find the driver. Certainly the shell script you have written is very useful. However I am still stuck. Sorry.

**Comment by MarionLegrandLogrami** [comment]
Hi @boshek - thank you for this detailed feedback on the problems encountered during installation. Depending on the error message, the problem is that you have installed an odbc driver that does not rely on the same architecture than the one used by R (you have a 64-bit R version and a 32-bit ODBC driver). There are 2 solutions:
1. Install the 64 bit odbc driver. The best way is to take the latest version of the driver regardless of the version of postgresql (I modified this instruction in the documentation). You can [download it here](https://ftp.postgresql.org/pub/odbc/versions/msi/psqlodbc_11_00_0000-x64.zip)
2. open 32-bit R and keep your 32-bit odbc driver

Regarding your first suggestion (include everything in the package), we unfortunately can not follow it because for the package to be accepted by CRAN, it is necessary that the data does not exceed 5MB. 

The compressed database is already 8 MB. If we try to include it in the package, we may not be able to submit it to CRAN.
![image](https://user-images.githubusercontent.com/26621584/57627718-d0331c00-7598-11e9-88dd-e4a48957d213.png)
(read on the [CRAN Repository Policy pages](https://cran.r-project.org/web/packages/policies.html))

Your feedback on the quality of the documentation seems very important for us. Unfortunately, I am not sure I understand the suggestion you made to improve the documentation. At this point, I have separated the odbc driver installation detail in a specific part of the documentation (part 6. More help), leaving only in the "Requirement" part the indication that you need to have an odbc driver and indicating a link to the specific part of the document in which I explain how to make this installation. You propose to separate this part. You mean to extract it completely from the documentation to put it in another help file? If you could give me some details I will be very grateful

**Comment by MarionLegrandLogrami** [comment]
Hello @boshek - Did you manage to install the 64-bit postgresql driver or did you try the 32-bit R version compatible with the postgresql driver that you already installed?
Please, let me know if you have any problem !

**Comment by danielskatz** [comment]
👋 @boshek - any news on this review?

**Comment by boshek** [comment]
@danielskatz sorry Daniel I haven't been able to prioritize this. I am very much still interested but my foreseable calendar is pretty full. Thank you for pinging me. I will endeavour to re-test this.

**Comment by kyleniemeyer** [comment]
@boshek can you let us know when you might be able to complete your review? It's ok if you need more time, but a target date would be helpful.

**Comment by MarionLegrandLogrami** [comment]
@kyleniemeyer - Dear Editor, we are very grateful for the time spent by our reviewers on the submission of our package to JOSS and for their comments. Thanks to them we made lot of progress ! I was planning to use this paper as part of my Phd Thesis for which time is running out as we are waiting to see our package accepted by JOSS. This is a package that has passed all R CRAN checks and is already used by many organizations on a daily basis, and we are freely and actively advising and helping end users, mostly people doing fish migration monitoring in the field. To be fully functional, the package requires a database connection which makes the installation more complex. We offered our reviewer to help him install the various necessary programs and also developed a specific module to help him install postgres and set up the connection. To date we have trouble understanding where he got troubles in the installation. So we are coming back to you to see if it could be possible to speed up the validation process of our package in JOSS ? Again we are at your disposal and that of the reviewer to do anything to help you. Thanking you in advance for your help.

**Comment by arfon** [comment]
@MarionLegrandLogrami - I've just emailed @boshek to ask if they can come and take another look at this review soon.

**Comment by boshek** [comment]
Ok I will commit to having this finished by the end of August 9th. Sorry to have languished so long here. I still think that too much complexity is being offloaded to the user here rather than abstracting that away.

**Comment by boshek** [comment]
I am actively working on this but the end of August 9th might not be possible. I will update my status at the end of August 9th.

**Comment by boshek** [comment]
This is an update to this process. 

## Using the stacomi package

I was running into issues because of leftover installations and other uses of a local PostgreSQL server. Additionally I was running into issues because I had installed the ODBC driver via the StackBuilder. That installed the driver in a different place than manually installing like the instructions said. **I would recommend having your installation script check both locations for the driver.** 

That said I have been able to only go a little further in the process in getting the package/database up and running. I am able to successfully bring up the UI using `stacomi()`. I do get this error however:

```

Loading required package: stacomirtools
Loading required package: RODBC
> stacomi()
C:/program files/stacomi/calcmig.csv does not exist, switching to defaut package fileLoading required package: gWidgetsRGtk2
Loading required package: RGtk2
Loading required package: gWidgets
Loading required package: cairoDevice
```
That said everything proceed in the sense that the dashboard appears. However I am not able to make it do anything (generate figures or tables) and I am assuming that the connection to the database isn't functioning properly. My first thought was to manually connect to the database in R and check if any actual data is there:

``` r
library(DBI)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(dbplyr)
#> 
#> Attaching package: 'dbplyr'
#> The following objects are masked from 'package:dplyr':
#> 
#>     ident, sql

con <- dbConnect(RPostgres::Postgres(),dbname = 'bd_contmig_nat', 
                 host = 'localhost',
                 port = 5432, 
                 user = 'postgres',
                 password = 'foo')

dbListTables(con)
#> [1] "chevauchement"

tbl(con, in_schema("iav", "t_bilanmigrationjournalier_bjo_bjo_identifiant_seq"))
#> # Source:   table<iav.t_bilanmigrationjournalier_bjo_bjo_identifiant_seq>
#> #   [?? x 3]
#> # Database: postgres [postgres@localhost:5432/bd_contmig_nat]
#>   last_value log_cnt is_called
#>   <int64>    <int64> <lgl>    
#> 1 351677     0       TRUE
```

<sup>Created on 2019-08-09 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

I am unclear that this output is telling me and whether it is helpful at all. However at this point this is as far as I could get. I thought to try installing the dev version but got some GTK errors which I am unable to debug at this point:

> Warning: Failed to load RGtk2 dynamic library, attempting to install it.
Please install GTK+ from http://ftp.gnome.org/pub/gnome/binaries/win32/gtk+/2.22/gtk+-bundle_2.22.1-20101227_win32.zip
If the package still does not load, please ensure that GTK+ is installed and that it is on your PATH environment variable
IN ANY CASE, RESTART R BEFORE TRYING TO LOAD THE PACKAGE AGAIN
Error in inDL(x, as.logical(local), as.logical(now), ...) : 
  unable to load shared object 'C:/Users/samal/Documents/R/win-library/3.6/RGtk2/libs/i386/RGtk2.dll':
  LoadLibrary failure:  The specified module could not be found.

## Some Other Considerations

### Installation Script
I still think that the installation script could be written completely in R avoiding the clunky step of downloading the installation file and running it. I think that you have misunderstood my meaning here. I don't want to run the installation script with R. I think it should be written in R itself so that a user can debug in the language they are most likely to be familiar with - R. This would provide a unified interface for users. However, I think given the length of this process and admitting that reasonable people can disagree on this point, this is not a point that is worth holding up acceptance of this review particularly since my more material concerns have been ameliorated by the process that the installation script enacts and very thorough documentation. 

### Shiny
After finally being able to generate the UI for this process I can't help but ask why you aren't doing this in Shiny. This comment is only relevant to this review because it would great ease a user experience (mine being documented here) taking the target audience for a dashboard (non R users) and taking them completely out of R. Without even considering the much improved UI experience, Shiny would be far easier for you to deploy. For a future I would strongly recommend using Shiny for the functionality currently being filled for GTK+.

**Comment by arfon** [comment]
This review has been going on for a very long time now. I'm very grateful to both @steffilazerte and @boshek for their work here - I believe the `stacomiR` package is in substantially better shape because of their time and efforts.

While I believe there are still some challenges with this software, I'm going to make the editorial decision at this point to accept this submission into JOSS.

@MarionLegrandLogrami - At this point could you make a new release of this software that includes the changes that have resulted from this review. Then, please make an archive of the software in Zenodo/figshare/other service and update this thread with the DOI of the archive? For the Zenodo/figshare archive, please make sure that:

- The title of the archive is the same as the JOSS paper title
- That the authors of the archive are the same as the JOSS paper authors

I can then move forward with accepting the submission.

**Comment by arfon** [comment]
@whedon add @boshek as reviewer

**Comment by whedon** [comment]
OK, @boshek is now a reviewer

**Comment by arfon** [comment]
> @whedon add @boshek as reviewer

@boshek - this is just to make sure you're listed at the top of this review thread. There's nothing new to do here.

**Comment by MarionLegrandLogrami** [comment]
Thank you very much for accepting this submission in JOSS and @steffilazerte and @boshek - thanks again for your time and advice!

@arfon - Here's the DOI (for the archive of the package):
[http://doi.org/10.5281/zenodo.3371259](http://doi.org/10.5281/zenodo.3371259)

**Comment by arfon** [comment]
@whedon set 10.5281/zenodo.3371259 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3371259" target="_blank">10.5281/zenodo.3371259</a> is the archive.

**Comment by arfon** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/907

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/907, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by arfon** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1163447176552493058) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/908
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.00791](https://doi.org/10.21105/joss.00791)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by arfon** [comment]
@steffilazerte @boshek, many thanks for your reviews here.

@MarionLegrandLogrami - your paper is now accepted into JOSS :zap::rocket::boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.00791/status.svg)](https://doi.org/10.21105/joss.00791)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.00791">
  <img src="https://joss.theoj.org/papers/10.21105/joss.00791/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.00791/status.svg
   :target: https://doi.org/10.21105/joss.00791
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.00791/status.svg)](https://doi.org/10.21105/joss.00791)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.salsalabs.org/donate-to-joss
